### ISSUE TITLE
Subchannel sharing does not seem to be working

### ISSUE THREAD

**Issue by evildour** [grpc/grpc#15207]
### Should this be an issue in the gRPC issue tracker?
Yes

### What version of gRPC and what language are you using?
C# - Grpc.Core 1.11.0
 
### What operating system (Linux, Windows, …) and version?
Windows 10 - 10.0.14393
 
### What runtime / compiler are you using (e.g. python version or version of gcc)
.NET Core 2.0

'dotnet --info' output:
```
.NET Command Line Tools (2.1.100)

Product Information:
 Version:            2.1.100
 Commit SHA-1 hash:  b9e74c6520

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.14393
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\2.1.100\

Microsoft .NET Core Shared Framework Host

  Version  : 2.0.5
  Build    : 17373eb129b3b05aa18ece963f8795d65ef8ea54
```
 
### What did you do?
Ran the following code in a console application with appropriate credentials:
```csharp
Environment.SetEnvironmentVariable("GRPC_TRACE", "subchannel");
Environment.SetEnvironmentVariable("GRPC_VERBOSITY", "DEBUG");

var channels = new List<Channel>();
for (int i = 0; i < 100; i++)
{
    channels.Add(new Channel("bigtable.googleapis.com", 443, channelCredentials));
}
foreach (var channel in channels)
{
    await channel.ConnectAsync();
}
Console.WriteLine("Channels connected, press enter to close them");
Console.ReadLine();
Console.WriteLine("Closing channels");
foreach (var channel in channels)
{
    await channel.ShutdownAsync();
}
```
 
### What did you expect to see?
I was under the impression that subchannels would be shared for channels with the same endpoint and options. It appears that would be the case here for all the channels.
 
### What did you see instead?
I see exactly 100 log messages that say something like `New connected subchannel at 000001E9587DD850 for subchannel 000001E958653B40`, meaning 1 subchannel has been used for each channel created. This can also be verified by running netstat before pressing the enter key and seeing 100 connections in the ESTABLISHED state. No sharing is happening. 

### Anything else we should know about your project / environment?
Nothing

**Comment by evildour** [comment]
Any update on this?

**Comment by AspirinSJL** [comment]
Sorry, I haven't worked on this yet. I will try to reproduce it in C++ tomorrow.

**Comment by AspirinSJL** [comment]
I am unable to reproduce it in C++. 

I added 20 same test channels in the end2end_test like the following:
```
diff --git a/test/cpp/end2end/end2end_test.cc b/test/cpp/end2end/end2end_test.cc
index 60238e930d..fc38d6767a 100644
--- a/test/cpp/end2end/end2end_test.cc
+++ b/test/cpp/end2end/end2end_test.cc
@@ -292,8 +292,13 @@ class End2endTest : public ::testing::TestWithParam<TestScenario> {
     args.SetString(GRPC_ARG_SECONDARY_USER_AGENT_STRING, "end2end_test");
 
     if (!GetParam().inproc) {
+      gpr_log(GPR_INFO, "Creating channels");
       channel_ =
           CreateCustomChannel(server_address_.str(), channel_creds, args);
+      for (size_t i = 0; i < 20; ++i) {
+        test_channels_.push_back(CreateCustomChannel(server_address_.str(), channel_creds, args));
+        gpr_log(GPR_INFO, "channel %lu created", i);
+      }
     } else {
       channel_ = server_->InProcessChannel(args);
     }
@@ -324,6 +329,7 @@ class End2endTest : public ::testing::TestWithParam<TestScenario> {
 
   bool is_server_started_;
   std::shared_ptr<Channel> channel_;
+  std::vector<std::shared_ptr<Channel>> test_channels_;
   std::unique_ptr<grpc::testing::EchoTestService::Stub> stub_;
   std::unique_ptr<Server> server_;
   std::unique_ptr<Server> proxy_server_;
```

The log after running `GRPC_VERBOSITY=DEBUG bins/opt/end2end_test --gtest_filter=End2endServerTryCancel/End2endServerTryCancelTest.RequestEchoServerCancel/1
` shows
```
Note: Google Test filter = End2endServerTryCancel/End2endServerTryCancelTest.RequestEchoServerCancel/1
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from End2endServerTryCancel/End2endServerTryCancelTest
[ RUN      ] End2endServerTryCancel/End2endServerTryCancelTest.RequestEchoServerCancel/1
D0521 19:00:13.913177842  174102 end2end_test.cc:216]        TestScenario{use_proxy=false, inproc=false, credentials='INSECURE_CREDENTIALS'}
D0521 19:00:13.913242835  174102 ev_posix.cc:145]            Using polling engine: epollex
D0521 19:00:13.913257351  174102 dns_resolver.cc:339]        Using native dns resolver
D0521 19:00:13.914543904  174102 ev_posix.cc:145]            Using polling engine: epollex
D0521 19:00:13.914562304  174102 dns_resolver.cc:339]        Using native dns resolver
I0521 19:00:13.914593831  174102 server_builder.cc:269]      Synchronous server. Num CQs: 4, Min pollers: 1, Max Pollers: 2, CQ timeout (msec): 10
I0521 19:00:13.914798923  174102 end2end_test.cc:295]        Creating channels
I0521 19:00:13.914852294  174102 end2end_test.cc:300]        channel 0 created
I0521 19:00:13.914870227  174102 end2end_test.cc:300]        channel 1 created
I0521 19:00:13.914885427  174102 end2end_test.cc:300]        channel 2 created
I0521 19:00:13.914900406  174102 end2end_test.cc:300]        channel 3 created
I0521 19:00:13.914915374  174102 end2end_test.cc:300]        channel 4 created
I0521 19:00:13.914930912  174102 end2end_test.cc:300]        channel 5 created
I0521 19:00:13.914946100  174102 end2end_test.cc:300]        channel 6 created
I0521 19:00:13.914963170  174102 end2end_test.cc:300]        channel 7 created
I0521 19:00:13.914983775  174102 end2end_test.cc:300]        channel 8 created
I0521 19:00:13.914998949  174102 end2end_test.cc:300]        channel 9 created
I0521 19:00:13.915015288  174102 end2end_test.cc:300]        channel 10 created
I0521 19:00:13.915029863  174102 end2end_test.cc:300]        channel 11 created
I0521 19:00:13.915046025  174102 end2end_test.cc:300]        channel 12 created
I0521 19:00:13.915060677  174102 end2end_test.cc:300]        channel 13 created
I0521 19:00:13.915076663  174102 end2end_test.cc:300]        channel 14 created
I0521 19:00:13.915091193  174102 end2end_test.cc:300]        channel 15 created
I0521 19:00:13.915109073  174102 end2end_test.cc:300]        channel 16 created
I0521 19:00:13.915123957  174102 end2end_test.cc:300]        channel 17 created
I0521 19:00:13.915140114  174102 end2end_test.cc:300]        channel 18 created
I0521 19:00:13.915154498  174102 end2end_test.cc:300]        channel 19 created
D0521 19:00:13.915184154  174102 dns_resolver.cc:280]        Start resolving.
I0521 19:00:13.915372517  174306 subchannel.cc:608]          New connected subchannel at 0x7f87f80065f0 for subchannel 0x7f87f80055f0
I0521 19:00:13.915551152  174308 test_service_impl.cc:342]   server_try_cancel : 1
I0521 19:00:13.915588618  174308 test_service_impl.cc:351]   Server called TryCancel() to cancel the request
[       OK ] End2endServerTryCancel/End2endServerTryCancelTest.RequestEchoServerCancel/1 (3 ms)
[----------] 1 test from End2endServerTryCancel/End2endServerTryCancelTest (3 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (3 ms total)
[  PASSED  ] 1 test.
D0521 19:00:13.916826812  174102 ev_posix.cc:145]            Using polling engine: epollex
D0521 19:00:13.916842111  174102 dns_resolver.cc:339]        Using native dns resolver
```

There is only one `New connected subchannel at 0x7f87f80065f0 for subchannel 0x7f87f80055f0` log.

Maybe a bug in C#? Re-assigning to @jtattermusch for further triaging.

**Comment by evildour** [comment]
Any update on this @jtattermusch?

**Comment by evildour** [comment]
Ping @jtattermusch

**Comment by jtattermusch** [comment]
I was able to reproduce the problem and I also have a theory why subchannel sharing doesn't work in this case.

A few things to know:
- in C core, for a subchannel to be reused, the new channel must created with channel arguments (ChannelOption in C#) that are equal (= same names and values) as channel arguments of an existing channel  (see https://github.com/grpc/grpc/blob/af89e8c00e796f3398b09b7daed693df2b14da56/src/core/ext/filters/client_channel/subchannel.cc#L330)
- internally, the ChannelCredentials are represented as a channel argument (= basically a pointer to the native ChannelCredentials object), so the channels must be created with the same credentials object for channel credentials to kick in.

With the current implementation in C#, the managed credentials (such as SslCredentials objects) are converted to a native object each time a channel gets created:
https://github.com/grpc/grpc/blob/af89e8c00e796f3398b09b7daed693df2b14da56/src/csharp/Grpc.Core/Channel.cs#L75
As a result, if one creates 2 channels from the same managed "ChannelCredentials" object, they will end up created with 2 different instances of native credential object, which breaks the subchannel sharing.

The idea of the fix is simple (we need to create channel using the same nativeCredentials object), but doing that will complicate management of nativeCredential's lifetime, so I will need to put some thought into it.

As a side note, if ChannelCredentials.Insecure is used, the subchannel sharing works fine (because no "credentials" pointer is passed to channels at creation time).

**Comment by jtattermusch** [comment]
Should be fixed by #16438 (which should be available in the upcoming 1.15.0 release).
